File: app/src/core/model/capital/capital.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import type { DeepReadonly } from "es-toolkit/types";
import type { Capital } from "@tsukilabs/nil-bindings";
import { PublicCapitalImpl } from "@/core/model/capital/public-capital";

export class CapitalImpl extends PublicCapitalImpl implements DeepReadonly<Capital> {
  private constructor(capital: Capital) {
    super(capital);
  }

  public static override create(capital: Capital) {
    if (capital instanceof CapitalImpl) {
      return capital;
    }

    return new CapitalImpl(capital);
  }
}

File: crates/nil-client/src/client/capital.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use super::Client;
use crate::error::Result;
use crate::http;
use nil_payload::request::capital::*;
use nil_payload::response::capital::*;

impl Client {
  /// Endpoint: `PUT /get-city-limit`
  pub async fn get_city_limit(&self, req: GetCityLimitRequest) -> Result<GetCityLimitResponse> {
    http::json_put("get-city-limit")
      .body(req)
      .server(self.server)
      .maybe_authorization(self.authorization.as_ref())
      .circuit_breaker(self.circuit_breaker())
      .retry(&self.retry)
      .user_agent(&self.user_agent)
      .send()
      .await
  }
}
